### PR TITLE
Remove feature

### PR DESCRIPTION
**Pull request by amitgilad3**

### CODE REVIEW

File: src/cli/commands/private-cmds/_list-cmd.js [hunk 1]
```diff
@@ -8,10 +8,10 @@ export default class List extends Command {
   private = true;
   description = 'list scope components';
   alias = '';
-  opts = [];
+  opts = [['a', 'all', 'show all components including soft-delete']];
```
Comment (GiladShoham) [inline src/cli/commands/private-cmds/_list-cmd.js]: remove the all from list (and from show)

File: src/cli/commands/private-cmds/scope-config-cmd.js [hunk 2]
```diff
@@ -86,6 +86,6 @@ class ScopeConfigDel extends Command {
   }
 
   report(conf: { [string]: string }): string {
-    return 'deleted successfully';
+    return 'deprecated successfully';
```
Comment (GiladShoham) [inline src/cli/commands/private-cmds/scope-config-cmd.js]: change to deleted

File: src/api/scope/lib/scope-list.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // @flow
 import { loadScope } from '../../../scope';
 
-export default function list(path: string): Promise<any> {
-  return loadScope(path).then(scope => scope.listStage());
+export default function list(path: string, all: boolean): Promise<any> {
```
Comment (GiladShoham) [inline src/api/scope/lib/scope-list.js]: remove the all, it's no longer needed

File: src/cli/commands/private-cmds/_list-cmd.js [hunk 4]
```diff
@@ -10,8 +10,8 @@ export default class List extends Command {
   alias = '';
   opts = [];
 
-  action([path]: [string, string]): Promise<any> {
-    return scopeList(fromBase64(path)).then(components => components.map(c => c.toString()));
+  action([path, all]: [string, string]): Promise<any> {
```
Comment (GiladShoham) [inline src/cli/commands/private-cmds/_list-cmd.js]: remove tha all

File: src/cli/commands/public-cmds/config-cmd.js [hunk 5]
```diff
@@ -84,6 +84,6 @@ class ConfigDel extends Command {
   }
 
   report(conf: { [string]: string }): string {
-    return 'deleted successfully';
+    return 'deprecated successfully';
```
Comment (GiladShoham) [inline src/cli/commands/public-cmds/config-cmd.js]: replace all mistake, restore the deleted

File: src/cli/commands/public-cmds/remove-cmd.js [hunk 6]
```diff
@@ -1,26 +1,49 @@
...
+  ];
 
-//   action([name]: [string], opts: any): Promise<any> {
-//     return remove(name, opts)
-//     .then(() => ({
-//       name,
-//       inline: opts.inline
-//     }));
-//   }
+  action([ids]: [string], { force = false, remote = false }: { force: Boolean, remote: Boolean }): Promise<any> {
```
Comment (GiladShoham) [inline src/cli/commands/public-cmds/remove-cmd.js]: boolean should be with lower b

File: src/cli/commands/public-cmds/scope-config-cmd.js [hunk 7]
```diff
@@ -86,6 +86,6 @@ class ScopeConfigDel extends Command {
   }
 
   report(conf: { [string]: string }): string {
-    return 'deleted successfully';
+    return 'deprecated successfully';
```
Comment (GiladShoham) [inline src/cli/commands/public-cmds/scope-config-cmd.js]: should be deleted

File: src/scope/models/component.js [hunk 8]
```diff
@@ -23,7 +23,8 @@ export type ComponentProps = {
   box?: string,
   name: string,
   versions?: { [number]: Ref },
-  lang?: string
+  lang?: string,
+  deprecated: Boolean
```
Comment (GiladShoham) [inline src/scope/models/component.js]: Boolean -> boolean

File: src/scope/network/ssh/ssh.js [hunk 9]
```diff
@@ -160,11 +177,10 @@ export default class SSH implements Network {
     return this.exec('_list').then((str: string) => {
       const { payload, headers } = this._unpack(str);
       checkVersionCompatibility(headers.version);
-      return rejectNils(
-        payload.map((c) => {
-          return c ? ConsumerComponent.fromString(c) : null;
-        })
-      );
+      const consumerComponents = payload.map((c) => {
```
Comment (GiladShoham) [inline src/scope/network/ssh/ssh.js]: Why did you change it?

File: src/scope/scope.js [hunk 10]
```diff
@@ -538,6 +538,91 @@ export default class Scope {
...
+    return component.id();
+  }
+
+  /**
+   * findDependentBits
+   * foreach component in array find the componnet that uses that component
+   */
+  async findDependentBits(bitIds: Array<BitId>): Promise<Array<object>> {
+    const stagedComponents = await this.listStage();
+    return new Promise((resolve) => {
```
Comment (GiladShoham) [inline src/scope/scope.js]: Don't use this form for return a promise.
it's better to just return 
```
Promise.resolve(dependentBits)
```
in the end

File: src/scope/scope.js [hunk 11]
```diff
@@ -538,6 +538,91 @@ export default class Scope {
...
+      const dependentBits = {};
+      bitIds.forEach((bitId) => {
```
Comment (GiladShoham) [inline src/scope/scope.js]: You can write this logic in a much better way using lodash/ramda but it's not that important.

File: src/scope/scope.js [hunk 12]
```diff
@@ -538,6 +538,91 @@ export default class Scope {
     return Promise.all(versionDependenciesArr.map(versionDependencies => versionDependencies.toConsumer(this.objects)));
   }
 
+  /**
+   * Remove or deprecate single component
+   * @removeComponent - boolean - true if you want to remove component
+   */
+  async removeSingle(bitId: BitId, removeComponent: boolean): Promise<string> {
```
Comment (GiladShoham) [inline src/scope/scope.js]: split this to 2 different functions remove / deprecate

File: src/scope/scope.js [hunk 13]
```diff
@@ -538,6 +538,91 @@ export default class Scope {
...
+    });
+  }
+
+  /**
+   * split bit array to found and missing components (incase user misspelled id)
+   */
+  async filterFoundAndMissingComponents(bitIds: Array<BitId>) {
+    const missingComponents = [];
+    const foundComponents = [];
+    const resultP = bitIds.map(async (id) => {
```
Comment (GiladShoham) [inline src/scope/scope.js]: better to use groupBy from lodash/ramda

File: src/scope/scope.js (hunk 10 above)
Comment (amitgilad3) [reply src/scope/scope.js]: fixed

File: src/scope/scope.js (hunk 12 above)
Comment (amitgilad3) [reply src/scope/scope.js]: done

File: src/scope/network/ssh/ssh.js (hunk 9 above)
Comment (amitgilad3) [reply src/scope/network/ssh/ssh.js]: my mistake, this was the change for showing deprecate component with a flag,
i reverted the change

File: src/scope/models/component.js (hunk 8 above)
Comment (amitgilad3) [reply src/scope/models/component.js]: done

File: src/api/consumer/lib/deprecate.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+import { loadConsumer } from '../../../consumer';
+import { BitId } from '../../../bit-id';
+
+export default (async function deprecate({ ids, remote }: { ids: string[], remote: boolean }): Promise<any> {
+  const bitIds = ids.map(bitId => BitId.parse(bitId));
+  const consumer = await loadConsumer();
+  if (remote) {
+    const groupedBitsByScope = groupArray(bitIds, 'scope');
+    const remotes = await consumer.scope.remotes();
+    const deprecateP = Object.keys(groupedBitsByScope).map(async (key) => {
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply src/api/consumer/lib/deprecate.js]: key is the scopename ,i changed the name

File: src/api/consumer/lib/deprecate.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
+/** @flow */
+import groupArray from 'group-array';
+import { loadConsumer } from '../../../consumer';
+import { BitId } from '../../../bit-id';
+
+export default (async function deprecate({ ids, remote }: { ids: string[], remote: boolean }): Promise<any> {
+  const bitIds = ids.map(bitId => BitId.parse(bitId));
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply src/api/consumer/lib/deprecate.js]: moved

File: e2e/commands/show.e2e.js [hunk 16]
```diff
@@ -221,6 +233,35 @@ describe('bit show command', function () {
...
+      helper.commitComponentBarFoo();
+      helper.exportAllComponents();
+    });
+    it('should show deprecated component', () => {
+      helper.deprecateComponent(`${helper.remoteScope}/bar/foo`, '-r');
+      output = JSON.parse(helper.runCmd(`bit show ${helper.remoteScope}/bar/foo -j`));
+      expect(output).to.include({ deprecated: true });
+    });
+  });
+  describe('show not deprecated remote component', () => {
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply e2e/commands/show.e2e.js]: already have one

File: e2e/commands/show.e2e.js [hunk 17]
```diff
@@ -221,6 +233,35 @@ describe('bit show command', function () {
...
+  describe('show deprecated remote component', () => {
+    let output;
+    before(() => {
+      helper.setNewLocalAndRemoteScopes();
+      helper.createComponentBarFoo();
+      helper.addComponentBarFoo();
+      helper.commitComponentBarFoo();
+      helper.exportAllComponents();
+    });
+    it('should show deprecated component', () => {
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply e2e/commands/show.e2e.js]: done

File: e2e/commands/show.e2e.js [hunk 18]
```diff
@@ -36,6 +36,19 @@ describe('bit show command', function () {
       helper.commitComponent('comp/comp');
     });
 
+    describe('show deprecated local component', () => {
+      let output;
+      it('should show deprecated component', () => {
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply e2e/commands/show.e2e.js]: done

File: e2e/commands/remove.e2e.js [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+      helper.reInitLocalScope();
+      helper.createComponentBarFoo();
+      helper.addComponentBarFoo();
+      helper.commitComponentBarFoo();
+    });
+    it('should remove component', () => {
+      const output = helper.removeComponent('bar/foo');
+      expect(output).to.contain.string('removed components: bar/foo');
+    });
+    it('should removed component from commited components back to new', () => {
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply e2e/commands/remove.e2e.js]: done

File: e2e/commands/remove.e2e.js [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+    before(() => {
+      helper.setNewLocalAndRemoteScopes();
+      helper.createComponentBarFoo();
+      helper.addComponentBarFoo();
+      helper.commitComponentBarFoo();
+      helper.exportAllComponents();
+    });
+    it('should remove remote component', () => {
+      const output = helper.removeComponent(`${helper.remoteScope}/bar/foo`, '-r');
+      expect(output).to.contain.string(`removed components: ${helper.remoteScope}/bar/foo`);
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply e2e/commands/remove.e2e.js]: i added tests for this in show and list e2e files

File: src/api/scope/lib/deprecate.js [hunk 21]
```diff
@@ -0,0 +1,10 @@
+// @flow
+import { loadScope } from '../../../scope';
+import BitId from '../../../bit-id/bit-id';
+
+export default function remove({ path, bitIds }): Promise<string[]> {
```
Comment (GiladShoham) [inline src/api/scope/lib/deprecate.js]: the function name should be deprecate i guess

File: src/cli/commands/private-cmds/_delete-cmd.js [hunk 22]
```diff
@@ -0,0 +1,21 @@
+/** @flow */
+import Command from '../../command';
+import { remove } from '../../../api/scope';
+import { fromBase64, unpackCommand, packCommand, buildCommandMessage } from '../../../utils';
+
+export default class DeprecatePrivate extends Command {
```
Comment (GiladShoham) [inline src/cli/commands/private-cmds/_delete-cmd.js]: Class name should be Delete
